text/3735-field-projections.md

```diff
@@ -0,0 +1,1745 @@
+}
+```
+
+The `Projectable` trait is what tells the compiler which types' fields to consider for projecting.
+When you write `a->b`, then `a` has to implement `Projectable` in order for the compiler to know
+which type to look up the field `b` for.
+
+The actual projection operation is governed by `Project<F>`:
+
+```rust
+impl<'a, T, F> Project<F> for DoubleRef<'a, T>
+where
+    // Since we want to project both of our references, we want the field to be aligned.
+    F: UnalignedField<Base = T> + Field,
+{
+    type Output = DoubleRef<'a, F::Type>;
+
+    fn project(self) -> Self::Output {
+        DoubleRef {
+            first: self.first.project(),
+            second: self.second.project(),
+        }
+    }
+}
+```
+
+The `Project<F>` trait governs projections for fields represented by the field type `F`. These field
+types are generated for all[^3] fields of all structs. A field type always implements the
+`UnalignedField` trait:
+
+[^3]: Almost all, fields that don't have a size, are not included.
+
+```rust
+pub unsafe trait UnalignedField {
+    type Base: ?Sized;
+    type Type: ?Sized;
+    const OFFSET: usize;
+}
+```
+
+`Base` is set to the parent struct containing the field, `Type` is set to the type of the field
+itself and `OFFSET` is the offset in bytes as returned by `offset_of!`.
+
+
+With the above projections in place, users can write the following code:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+
+let x: &Foo = &Foo { bar: 42, baz: 43 };
+let y: &Foo = &Foo { bar: 24, baz: 25 };
+let d: DoubleRef<'_, Foo> = DoubleRef::new(x, y);
+let bars: DoubleRef<'_, i32> = d->bar;
+assert_eq!((42, 24), bars.read());
+```
+
+#### Containers
+
+The other kind of projections are that of *containers*, for example `UnsafeCell<T>` or
+`MaybeUninit<T>`. They are governed by the `#[projecting]` attribute.
+
+If we want to combine the two containers from the previous sections, we can do it in the following
+way:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Opaque<T> {
+    inner: UnsafeCell<MaybeUninit<T>>,
+}
+
+impl<T> Opaque<T> {
+    pub fn new(value: T) -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::new(value)) }
+    }
+
+    pub fn uninit() -> Self {
+        Self { inner: UnsafeCell::new(MaybeUninit::uninit()) }
+    }
+
+    pub fn write(&mut self, value: T) {
+        self.inner.get_mut().write(value);
+    }
+}
+```
+
+Now `&Opaque<T>` can represent a reference that points at data from other languages, such as C.
+
+The `#[projecting]` attribute changes the way the field types are generated for this type. Instead
+of having a field type representing the `inner` field, this type will "project" through to the type
+`T`, inheriting all fields that `T` has. So if we consider the type `Foo` from above:
+
+```rust
+struct Foo {
+    bar: i32,
+    baz: u32,
+}
+```
+
+Then there are two field types for `Opaque<Foo>`:
+- one representing `bar` with `Base = Opaque<Foo>` and `Type = Opaque<i32>`
```

Comment:
why not just use the field types for `Foo` directly, so the field type for `bar` impls `Field<Base = Foo, Type = i32>`?

maybe change the design like the following, so you don't need a magic `#[projecting]` attribute:

```rust
pub unsafe trait FieldType {} // maybe omit

/// Self is what used to be `Field::Base`
pub unsafe trait HasUnalignedField<F: FieldType> { // maybe leave `Has` out of name?
    type Type: ?Sized;

    const OFFSET: usize;
}

pub unsafe trait HasField<F: FieldType>: HasUnalignedField<F> {}

pub trait Projectable: Sized {
    type Inner: ?Sized;
}

pub trait Project<F: FieldType>: Projectable<Inner: UnalignedField<F>> {
    type Output;

    fn project(self) -> Self::Output;
}

unsafe impl<F: FieldType, T: HasUnalignedField<F, Type: Sized>> HasUnalignedField<F> for Cell<T> {
    type Type = Cell<<T as HasUnalignedField<F>>::Type>;
    const OFFSET: usize = <T as HasUnalignedField<F>>::OFFSET;
}

unsafe impl<F: FieldType, T: HasField<F, Type: Sized>> HasField<F> for Cell<T> {}
```

this allows implementing `HasUnalignedField`/`HasField` multiple times for the same field type, it also could allow:
```rust

struct MyWrapper<T> {
    other_field: u32,
    field: T,
}

unsafe impl<F: FieldType, T: HasUnalignedField<F, Type: Sized>> HasUnalignedField<F> for MyWrapper<T> {
    type Type = <T as HasUnalignedField<F>>::Type;
    const OFFSET: usize = <T as HasUnalignedField<F>>::OFFSET + offset_of!(Self, field);
}

unsafe impl<F: FieldType, T: HasField<F, Type: Sized>> HasField<F> for MyWrapper<T> {}
```

it also allows you to have a `HashMap` where the key is the `TypeId` of the field type and it works even if you access fields of `MyStruct` through `SomeRef<MyStruct>` or `SomeOtherType<MyStruct>`, since in both cases the field types are the same types -- the field types of `MyStruct`...

Using something like that `HashMap` is how I would probably implement field access for fayalite's `Expr<MyStruct>` if I were redesigning the API to use the projection operator introduced by this RFC.

Reply:
That looks very interesting!

How does the compiler find the right field type to return via `field_of!` though? Let's say I have the `MyWrapper` struct from above (it should probably be `#[repr(C)]`) as well as `Foo`:
```rust
struct Foo {
    other_field: i32,
    field: u32,
}

fn demo(x: &MyWrapper<Foo>) {
    let field = x->field; // which `field` is meant by this?
    let other_field = x->other_field; // same here
}
```

Reply:
> How does the compiler find the right field type to return via `field_of!` though? Let's say I have the `MyWrapper` struct from above (it should probably be `#[repr(C)]`) as well as `Foo`:

`#[repr(C)]` is unnecessary since we're using `offset_of!`. I didn't think of finding the right inner type...maybe we need `MyWrapper<T>: Projectable<Inner = T>` and the compiler follows the chain of `Projectable` impls kinda like deref?
> 
> ```rust
> struct Foo {
>     other_field: i32,
>     field: u32,
> }
> 
> fn demo(x: &MyWrapper<Foo>) {
>     let field = x->field; // which `field` is meant by this?
>     let other_field = x->other_field; // same here
> }
> ```

assuming `MyWrapper<Foo>: Projectable`, since MyWrapper opted-in to being projectable, those would resolve to `Foo`'s fields. alternatively they could be ambiguous.

Reply:
I'm proposing the desugaring would change so `a->b` desugars to something like:
```rust
type Base<...> = {
    let mut base: type = typeof(a);

    // only true if the impl is visible, it can be hidden due to generics or impl Trait
    while base: Projectable {
        base = base::Inner;
    }
    base
};
a.project(field_of!(Base, b) {})
```

Reply:
or maybe a better option is following `deref` coercion semantics where it stops at the first type with that field where the field is visible (due to `pub`).

Reply:
That sounds reasonable, though might lead to people having to do some visibility gymnastics:
```rust
mod my_wrapper_def {
    pub struct MyWrapper<T> {
        other_field: u32,
        field: T,
    }
    /* inherit all fields of `field: T`... */
}
pub use my_wrapper_def::MyWrapper;
pub mod foo_def {
    use super::*;
    pub struct Foo {
        bar: i32,
        baz: u32
    }

    impl MyWrapper<Foo> {
        pub fn set_bar(this: &mut MaybeUninit<Wrapper<Foo>>, value: i32) {
            let bar: &mut MaybeUninit<i32> = this->bar;
            bar.write(value);
        }
    }
}
```
Which probably is rather annoying... So while we should use your suggested algorithm, I think it's probably still very useful to not generate field types for certain fields via an attribute:
```rust
pub struct MyWrapper<T> {
    #[no_field_type_bikeshed]
    other_field: u32,
    #[no_field_type_bikeshed]
    field: T,
}
/* inherit all fields of `field: T`... */
pub struct Foo {
    bar: i32,
    baz: u32,
}

impl MyWrapper<Foo> {
    pub fn set_bar(this: &mut MaybeUninit<Wrapper<Foo>>, value: i32) {
        this->bar.write(value);
    }
}
```